feat: custom short link (back-end)

MaestroDagan opened the pull request:
closes #33 

## Summary
- if custom short link is provided it'll use that
- if custom short link is already used it'll respond with a 409 Conflict status code
- custom short link has basic validation
- updated docs
## Checklist
- [ ] fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ] tested locally.
- [x] No error nor warning in the console.

MaestroDagan committed:
feat: delete link endpoint

MaestroDagan committed:
feat: custom short link (back-end)

cherylli commented on server/src/controllers/links.controller.js:
```diff
@@ -1,44 +1,50 @@
+    return res
+      .status(409)
+      .json({ success: false, message: "short URL already exists" });
+  const randomString = req.body.short ?? (await getRandomString());
```
I would rename this variable to shortCode or something, since it's not completely random anymore.

MaestroDagan replied:
I've added all of the requested changes

cherylli reviewed (changes requested):
Could you also add 
1. basic validation of custom short links (min, max length, has to be valid characters) - right now, and empty string is accepted and saved in the database
2. edit the api documentation (swagger - docs.yaml) to reflect this change

MaestroDagan committed:
added requested changes

feat: requirements
fix: empty strings
chore: names

Length has to be between the minLength & maxLength
The link can only contain letters and -
If the shortUrl is empty it'll generate a randomString
The constant variable randomString has been changed to shortCode

cherylli committed:
Merge branch 'main' into feat-custom-short-link-(back-end)

cherylli commented:
Does index only consist of formatting change? If so, could you please remove from staged
Do you also mind reverting the formatting of existing code?
Thank you

MaestroDagan commented:
> 

I'll do that, thanks for your suggestions.
I think this is a really nice project to work on.

MaestroDagan committed:
Update index.js

MaestroDagan committed:
chore: removed formatting

MaestroDagan commented:
Removed almost all the formatting!

MaestroDagan committed:
Merge branch 'main' into feat-custom-short-link-(back-end)

MaestroDagan commented:
what happened?

MaestroDagan commented:
Because i renamed the branche te pr closed

MaestroDagan commented:
i reopened it, sorry for this.

cherylli committed:
Merge branch 'main' into feat-custom-short-link-(back-end)

cherylli committed:
minor fixes

cherylli commented:
hello again, could you please check if "“Allow edits from maintainers.”  is selected in this pr. I found some errors and would like to push to your branch

MaestroDagan commented:
Okay, I'll see

MaestroDagan commented:
Where can i find that setting? @cherylli

cherylli commented:
![image](https://user-images.githubusercontent.com/6191116/193593156-503b7bb2-bddc-4c0b-8c57-e2da6af8ba71.png)

MaestroDagan commented:
i dont think i can still change that

MaestroDagan commented:
but i could open a new PR with that checked, do you want me to do that? @cherylli

cherylli commented:
> but i could open a new PR with that checked, do you want me to do that? @cherylli

no please give me some time. I think I have messed up with the push

MaestroDagan commented:
Okay!

cherylli committed:
Update links.controller.js

cherylli committed:
Update docs.yaml

cherylli commented:
ok i think that should work
Just a few comments. It feels like you did not test the code as there were quite a few signifant errors
1. the regex didn't work, everything I tried was invalid
2. when a request was sent without "short", it gives an error because req.body.short doesn't exist, you should always check if a variable exist before using it or use optional chaining in some cases
3. some statuscode were inaccurate

When you do a pull request it's very important to check that everything is working

Let me do a final check to make sure my commit is working

cherylli committed:
Merge branch 'main' into feat-custom-short-link-(back-end)
